client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -122,26 +129,29 @@ private final class PoolManager[F[_], A <: Connection[F]](
 
               case Some(closedConn) =>
                 logger.debug(s"Evicting closed connection: $stats")
-                curTotal -= 1
                 decrConnection(key)
                 go()
 
-              case None
-                  if curTotal < maxTotal && allocated.getOrElse(key, 0) < getMaxConnections(key) =>
+              case None if numConnectionsCheckHolds(key) =>
                 logger.debug(s"Active connection not found. Creating new one. $stats")
                 createConnection(key, callback)
 
-              case None if idleQueue.get(key).exists(_.nonEmpty) =>
+              case None if curTotal == maxTotal =>
                 logger.debug(
                   s"No connections available for the desired key. Evicting oldest and creating a new connection: $stats")
-                curTotal -= 1
-                decrConnection(key)
-                getConnectionFromQueue(key).get.shutdown()
-                createConnection(key, callback)
+                val nonEmptyKeys = getNonEmptyKeys
+                if (nonEmptyKeys.nonEmpty) {
+                  val randKey = nonEmptyKeys.drop(Random.nextInt(nonEmptyKeys.size)).head
+                  idleQueues(randKey).dequeue().shutdown()
```

Comment:
How is this idleQueue getting cleaned up if it's the last for the key?  When the connection gets shut down and we call invalidate?

I think a List is not ideal for randomly jumping into the middle, even though it's fast to construct.  Maybe a Vector would be a better choice?

Reply:
Oh yeah, the keys that we get from the map are not in a vector. Optimizing the nonEmpty key collection I totally forgot that I am making a new collection so that could be anything ! 👅

Reply:
For the cleanup, we are calling decrConnection.